User Auth Providers (#1745)
neo4j/docs-operations: modules/ROOT/pages/authentication-authorization/auth-providers.adoc

:description: This section explains how to use Cypher to manage authentication and authorization at the user level using Cypher.

[role=label--new-5.24]
[[access-control-auth-providers]]
= User Auth Providers label:enterprise-edition[]

Authentication and authorization can be controlled on a user-level using Cypher by setting Auth Providers on users.

In order to make use of Auth Providers, you need to set the `dbms.security.require_local_user` configuration setting to `true`. This setting mandates that only users with a corresponding Auth Provider in the database can authenticate and authorize.

> Auth Providers give you a way to link externally-defined users (e.g. in a 3rd party ID provider like OIDC or LDAP) to the Neo4j internal User model. The internal model can define things such as roles (authorization), `SUSPENDED` status, `HOME DATABASE`, and metadata such as the unique displayed name of the user. For consistency you can also define `native` (password-based) auth using the Auth Provider syntax.

Reviewer: still valid, is now line 12

> Auth Providers give you a way to link externally-defined users (e.g. in a 3rd party ID provider like OIDC or LDAP) to the Neo4j internal User model. The internal model can define things such as roles (authorization), `SUSPENDED` status, `HOME DATABASE`, and metadata such as the unique displayed name of the user. For consistency you can also define `native` (password-based) auth using the Auth Provider syntax.

Author: thanks, done.

:description: This section explains how to use Cypher to manage authentication and authorization at the user level using Cypher.

> [role=label--new-5.24]

Author: for some reason this doesn't render properly @renetapopova

== Migrating to Auth Providers mode
If you have existing users in the database and you want to migrate to Auth Providers mode, you can use the `ALTER USER ... SET AUTH` command to attach Auth Provider(s) to an existing user. Until you change `dbms.security.require_local_user` to `true`, this will have no impact on the users' ability to authenticate and authorize as they always have done.

Once the process of adding Auth Providers to your users is completed, you can then set `dbms.security.require_local_user` to true and restart the DBMS to complete the migration. After this time, only users with a corresponding Auth Provider in the database will be able to authenticate and authorize.

[NOTE]
====
Existing users which were created using the original `CREATE USER ... SET PASSWORD` syntax, will implicitly have the native (username / password) Auth Provider, so that would not need to be added explicitly using `SET AUTH`.

> To be certain about which Auth Providers are attached to a user, you can use the xref:authentication-authorization/manage-users.adoc#access-control-list-users[`SHOW USERS WITH AUTH`] command to see the Auth Providers attached to a user.

Author: FYI @b33f

If there is a match, then the user is linked to the Neo4j user and authorized according to the DBMS security configuration settings that match the provider's name of the matching auth provider.

How the matching lookup is done depends on the type of provider.
For example:

* For an OIDC provider, the claim configured by xref:configuration/configuration-settings.adoc#config_dbms.security.oidc.-provider-.claims.username[`dbms.security.oidc.mysso.claims.username`] (default `sub`) is taken from the token and is used to look up an auth provider whose `ID` and `provider` properties match the `sub` and provider respectively of the OIDC provider.
* For an LDAP provider, the `dn` is used to look up an auth provider with a `provider` property of `ldap` and an `ID` property that matches the supplied `dn`.
* For the `native` (username/password) provider, the supplied username itself is used to look up the auth provider.

== Enabling user auth providers mode

To enable user auth providers mode, set the configuration setting xref:configuration/configuration-settings.adoc#config_dbms.security.require_local_user[`dbms.security.require_local_user`] to `true`.

> This setting ensures that users from the authentication provider must exist in the database before they can log in and be authorized.

Author: ```suggestion
This setting mandates that users with the relevant auth provider attached to them must exist in the database before they can authenticate and authorize with that auth provider.
```

> When the user authenticates, Neo4j searches for the user's authentication provider.
> If a matching authentication provider is found for that user, the user can log in and be authorized successfully.

Author: ```suggestion
When the user authenticates, Neo4j searches for a user with a matching authentication provider.
If a match is found, the user can log in and be authorized successfully.
```
